### PR TITLE
Retail coupon example: ClippedCouponCredential

### PR DESCRIPTION
**Pull request by ottonomy**
This is an interesting use case that would highlight the power of presentation of credentials at point of sale. The traditional coupon use case could become much more interesting for retailers and manufacturers when they could track additional (potentially anonymized) consumption patterns for products related to coupon promotion, clipping and redemption.

---

A retail coupon is traditionally a snippet of paper clipped out of a newspaper or promotion. Now consumers can clip Verifiable Credentials. Often digital coupons are represented with an image that includes a thick dashed border, to signal "cut here". To redeem, consumers might select a coupon from a retailer's app or website and obtain a `ClippedCouponCredential` from the retailer that can be redeemed at the point of sale. Each clipped coupon identifies a `CouponOffer`, which carries metadata defined by the manufacturer about a discount offer. An offer sets the value of a discount, the products to which it applies, and the conditions under which it can be redeemed. This information is sufficient to render a scannable [GS1](https://www.gs1us.org/upcs-barcodes-prefixes/additional-ways-to-identify-products/coupons) barcode that could provide a traditional paper-like presentation of the coupon where presentation of the VC would be impractical.

Example lifecycle of a AI (8112) coupon:
* Manufacturers define offer data 
* They store it in the relevant repository, the Coupon Bureau's Master Offer File, for the [AI (8112) process](https://help.thecouponbureau.org/docs/retailer-ai-8112-coupon-processing-requirement-overview). 
Offers are made available to Consumers, who "clip" coupons within shopping apps or websites. 
Customers make relevant purchases from retailers. Clipped coupons are presented. 
* The retailer validates coupons and their requirements and honors any discounts.
* Periodically, retailer submits redeemed Coupons to a retailer clearinghouse. 
* Retailer clearinghouse reimburses retailer for discounts provided to consumers and reconciles payments with manufacturers.

### CODE REVIEW

**Review by gannan08 — APPROVED** [review]
LGTM. Will merge down after the refactor.

File: credentials/retailCoupon/credential.json
Comment (dlongley) [inline credentials/retailCoupon/credential.json:11]: I think this example wouldn't work because of `image` and `url` are undefined properties (I think, I may have missed their definitions).

File: contexts/retailCoupon-context-v1.json
Comment (dlongley) [inline contexts/retailCoupon-context-v1.json:19]: There's a mistake in the context here where the closing bracket for `@context` is here and then `CouponOffer` is not nested inside the `@context` body.

**Review by dlongley — APPROVED** [review]
There are a couple of things required to fix up this example that I noted, but approving otherwise. Thanks!

File: credentials/retailCoupon/credential.json
Comment (ottonomy) [reply credentials/retailCoupon/credential.json]: Added 
```
    "image": {
      "@id": "https://schema.org/image"
    },
    "url": {
      "@id": "https://schema.org/url"
    },
```
at the top level of the context so it can apply on the VerifiableCredential or VerifiableCredential.issuer level.

**Comment by ottonomy** [comment]
@gannan08 @davidlehn refactored to follow pending guidelines in #15:
* renamed `retailCoupon` to `retail-coupon`
* renamed images to `image.ext` (and added additional PNG version as specified. 
* SVG is still present in repo to assist with generating new version PNGs in the future. Could also reference a SVG with support for RenderMethod if needed
* Vocabulary file added and attempt in context made to appropriately link terms to where this file will likely be hosted.

It would definitely be good to review with subject matter experts on the politics of coupons.
* Use of a databar in the image is probably an anti-pattern. Presentation of the digital credential itself would be better than generation of a scannable barcode. That would better comply with anti-fraud efforts in the coupon ecosystem. The barcode in the example image is nonsensical.
* Many of the terms inside the CouponOffer do not strictly need to be read by the retailer, because the retailer will take the `base_gs1` and perform a lookup from an authoritative repository to get up-to-date offer information before redeeming a coupon. It may be useful to provide them to enable broader rendering by other parties who would not do this lookup, but it may be deemed extraneous.
* Date formats within AI 8112 data are `MM/DD/YYYY`. This may be incompatible with  `xsd:dateTime` datatype defined in context.

**Comment by JSAssassin** [comment]
@ottonomy 
The vocab `index.html` file needs to be inside a `vocab` directory inside `contexts/retail-coupon` and then the vocab URLs will need to be updated.  https://github.com/credential-handler/vc-examples/pull/15/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R75-R78

**Comment by ottonomy** [comment]
@JSAssassin:
> @ottonomy The vocab `index.html` file needs to be inside a `vocab` directory inside `contexts/retail-coupon` and then the vocab URLs will need to be updated. 

Resolved in 4fdefc1

File: contexts/retail-coupon/vocab/index.html
Comment (davidlehn) [inline contexts/retail-coupon/vocab/index.html:407]: I see a handful of these closing angle brackets on another line.  Is that needed here?
```suggestion
      </pre>
```

File: contexts/retail-coupon/v1.json
Comment (davidlehn) [inline contexts/retail-coupon/v1.json:19]:
```suggestion
      "@id": "https://schema.org/image",
      "@type": "@id"
```

File: contexts/retail-coupon/v1.json
Comment (davidlehn) [inline contexts/retail-coupon/v1.json:22]:
```suggestion
      "@id": "https://schema.org/url",
      "@type": "@id"
```

File: contexts/retail-coupon/v1.json
Comment (davidlehn) [inline contexts/retail-coupon/v1.json:42]: `xsd` is not defined. Either define it somewhere or inline the full URL.  Similar for other uses here.
```suggestion
          "@type": "http://www.w3.org/2001/XMLSchema#dateTime"
```

File: contexts/retail-coupon/v1.json
Comment (davidlehn) [inline contexts/retail-coupon/v1.json:10]: All the places where it's only defining the id of a term can be simplified to one-liners.
```suggestion
    "ClippedCouponCredential": "https://contexts.vcplayground.org/examples/retail-coupon/vocab#ClippedCouponCredential",
```

**Review by davidlehn — CHANGES_REQUESTED** [review]

File: contexts/retail-coupon/v1.json
Comment (JSAssassin) [reply contexts/retail-coupon/v1.json:42]: if using the full URL we should use `https` instead if possible.

**Comment by JSAssassin** [comment]
The PR also requires a changelog entry in `CHANGELOG.md`. See [changelog guidelines](https://github.com/digitalbazaar/bedrock/blob/main/CONTRIBUTING.md#changelogs)

File: contexts/retail-coupon/v1.json
Comment (davidlehn) [reply contexts/retail-coupon/v1.json:42]: In this case I think we need to use `http`.  This is a situation of an "identifier" URL that has been long and widely used as `http` and tools may depend on that.  w3.org has all the HTTP things setup to redir and force `https` if you dereference it.  There are some other `http` ones here: https://www.w3.org/TR/json-ld/#conformance.  But of course, schema.org is an exception where they historically and unfortunately started with `http` and now using `https`.

Comment (davidlehn) [reply contexts/retail-coupon/v1.json:42]: The official namespace they use is listed here: https://www.w3.org/TR/xmlschema-2/#namespaces

File: contexts/retail-coupon/vocab/index.html
Comment (ottonomy) [reply contexts/retail-coupon/vocab/index.html:407]: Sometimes the auto-formatters do that in HTML when they're trying to preserve some intent about whitespace. Looks like it was somewhat confused this time.

**Review by gannan08 — APPROVED** [review]

**Comment by ottonomy** [comment]
There we challenges using the Rebase and Merge button here because of merged commits from `main`, so closing this one. @gannan08 this PR is replaced with #20.

**Pull request closed without merging**
